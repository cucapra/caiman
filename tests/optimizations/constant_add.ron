// ARGS: -t constant-add,basic-cse
(
	version : (0, 0, 1),
	program : (
		types : {0:I32, 1:U32},
		funclets : {
            0 : (
                // *definitely* can't constant fold this
                kind : MixedImplicit,
                input_types : [0, 0],
                output_types : [0],
                nodes : [
                    Phi(index : 0),
                    Phi(index : 1),
                    Binop(kind : Add, arg0: 0, arg1: 1)
                ],
                tail_edge : Return(return_values : [2]) 
            ),
            1 : (
                // can't constant fold this
                kind : MixedImplicit,
                input_types : [0],
                output_types : [0, 0],
                nodes : [
                    Phi(index : 0),
                    ConstantInteger(value: 10, type_id: 0),
                    Binop(kind : Add, arg0: 0, arg1: 1),
                    Binop(kind : Add, arg0: 1, arg1: 0)
                ],
                tail_edge : Return(return_values : [2, 3]) 
            ),
            2 : (
                // can't constant fold this
                kind : MixedImplicit,
                input_types : [1],
                output_types : [1, 1],
                nodes : [
                    Phi(index : 0),
                    ConstantUnsignedInteger(value: 10, type_id: 0),
                    Binop(kind : Add, arg0: 0, arg1: 1),
                    Binop(kind : Add, arg0: 1, arg1: 0)
                ],
                tail_edge : Return(return_values : [2, 3]) 
            ),
            3 : (
                // *can* constant fold this, kinda
                kind : MixedImplicit,
                input_types : [0],
                output_types : [0, 0],
                nodes : [
                    Phi(index : 0),
                    ConstantInteger(value: 0, type_id: 0),
                    Binop(kind : Add, arg0: 0, arg1: 1),
                    Binop(kind : Add, arg0: 1, arg1: 0)
                ],
                tail_edge : Return(return_values : [2, 3]) 
            ),
            4 : (
                // *can* constant fold this, kinda
                kind : MixedImplicit,
                input_types : [1],
                output_types : [1, 1],
                nodes : [
                    Phi(index : 0),
                    ConstantUnsignedInteger(value: 0, type_id: 0),
                    Binop(kind : Add, arg0: 0, arg1: 1),
                    Binop(kind : Add, arg0: 1, arg1: 0)
                ],
                tail_edge : Return(return_values : [2, 3]) 
            ),
            5 : (
                // ok time for REAL constant folding
                kind : MixedImplicit,
                input_types : [],
                output_types : [0],
                nodes : [
                    ConstantInteger(value: 17, type_id: 0),
                    ConstantInteger(value: -15, type_id: 0),
                    ConstantInteger(value: 0, type_id: 0),
                    ConstantInteger(value: 9, type_id: 0),
                    Binop(kind : Add, arg0: 0, arg1: 2), // = 17
                    Binop(kind : Add, arg0: 4, arg1: 1), // = 2
                    Binop(kind : Add, arg0: 3, arg1: 3), // = 18
                    Binop(kind : Add, arg0: 6, arg1: 5)  // = 20
                ],
                tail_edge : Return(return_values : [7]) 
            ),
            6 : (
                // like above, but unsigned
                kind : MixedImplicit,
                input_types : [],
                output_types : [1],
                nodes : [
                    ConstantUnsignedInteger(value: 17, type_id: 1),
                    ConstantUnsignedInteger(value: 2, type_id: 1),
                    ConstantUnsignedInteger(value: 0, type_id: 1),
                    ConstantUnsignedInteger(value: 9, type_id: 1),
                    Binop(kind : Add, arg0: 0, arg1: 2), // = 17
                    Binop(kind : Add, arg0: 4, arg1: 1), // = 19
                    Binop(kind : Add, arg0: 3, arg1: 3), // = 18
                    Binop(kind : Add, arg0: 6, arg1: 5)  // = 37
                ],
                tail_edge : Return(return_values : [7]) 
            ),
            7 : (
                // complex time :)
                kind : MixedImplicit,
                input_types : [1],
                output_types : [0],
                nodes : [
                    Phi(index : 0), // imagine this being an array pointer or something
                    ConstantUnsignedInteger(value: 81, type_id: 1), // and this is the offset
                    Binop(kind : Add, arg0: 0, arg1: 1), // so we get pointer to the element...
                    ConstantUnsignedInteger(value: 1, type_id: 1), 
                    Binop(kind : Add, arg0: 1, arg1: 3), 
                    Binop(kind : Add, arg0: 0, arg1: 4), // and pointer to the next element
                    CallExternalCpu(external_function_id: 0, arguments: [2, 5]), // and we swap em
                    ConstantInteger(value: 0, type_id: 0) // and return an exit code idk
                ],
                tail_edge : Return(return_values : [6]) 
            ),
        }
	)
)