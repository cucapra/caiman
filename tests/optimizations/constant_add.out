(
    version: (0, 0, 1),
    program: (
        types: {
            0: I32,
            1: U32,
        },
        funclets: {
            0: (
                kind: MixedImplicit,
                input_types: [
                    0,// [0]
                    0,
                ],
                output_types: [
                    0,
                ],
                nodes: [
                    Phi(
                        index: 1,
                    ),// [0]
                    Phi(
                        index: 0,
                    ),// [1]
                    Binop(
                        kind: Add,
                        arg0: 1,
                        arg1: 0,
                    ),
                ],
                tail_edge: Return(
                    return_values: [
                        2,
                    ],
                ),
                input_resource_states: [],
                output_resource_states: [],
                local_meta_variables: {},
            ),
            1: (
                kind: MixedImplicit,
                input_types: [
                    0,
                ],
                output_types: [
                    0,// [0]
                    0,
                ],
                nodes: [
                    ConstantInteger(
                        value: 10,
                        type_id: 0,
                    ),// [0]
                    Phi(
                        index: 0,
                    ),// [1]
                    Binop(
                        kind: Add,
                        arg0: 1,
                        arg1: 0,
                    ),
                ],
                tail_edge: Return(
                    return_values: [
                        2,// [0]
                        2,
                    ],
                ),
                input_resource_states: [],
                output_resource_states: [],
                local_meta_variables: {},
            ),
            2: (
                kind: MixedImplicit,
                input_types: [
                    1,
                ],
                output_types: [
                    1,// [0]
                    1,
                ],
                nodes: [
                    ConstantUnsignedInteger(
                        value: 10,
                        type_id: 0,
                    ),// [0]
                    Phi(
                        index: 0,
                    ),// [1]
                    Binop(
                        kind: Add,
                        arg0: 1,
                        arg1: 0,
                    ),
                ],
                tail_edge: Return(
                    return_values: [
                        2,// [0]
                        2,
                    ],
                ),
                input_resource_states: [],
                output_resource_states: [],
                local_meta_variables: {},
            ),
            3: (
                kind: MixedImplicit,
                input_types: [
                    0,
                ],
                output_types: [
                    0,// [0]
                    0,
                ],
                nodes: [
                    Phi(
                        index: 0,
                    ),
                ],
                tail_edge: Return(
                    return_values: [
                        0,// [0]
                        0,
                    ],
                ),
                input_resource_states: [],
                output_resource_states: [],
                local_meta_variables: {},
            ),
            4: (
                kind: MixedImplicit,
                input_types: [
                    1,
                ],
                output_types: [
                    1,// [0]
                    1,
                ],
                nodes: [
                    Phi(
                        index: 0,
                    ),
                ],
                tail_edge: Return(
                    return_values: [
                        0,// [0]
                        0,
                    ],
                ),
                input_resource_states: [],
                output_resource_states: [],
                local_meta_variables: {},
            ),
            5: (
                kind: MixedImplicit,
                input_types: [],
                output_types: [
                    0,
                ],
                nodes: [
                    ConstantInteger(
                        value: 20,
                        type_id: 0,
                    ),
                ],
                tail_edge: Return(
                    return_values: [
                        0,
                    ],
                ),
                input_resource_states: [],
                output_resource_states: [],
                local_meta_variables: {},
            ),
            6: (
                kind: MixedImplicit,
                input_types: [],
                output_types: [
                    1,
                ],
                nodes: [
                    ConstantUnsignedInteger(
                        value: 37,
                        type_id: 1,
                    ),
                ],
                tail_edge: Return(
                    return_values: [
                        0,
                    ],
                ),
                input_resource_states: [],
                output_resource_states: [],
                local_meta_variables: {},
            ),
            7: (
                kind: MixedImplicit,
                input_types: [
                    1,
                ],
                output_types: [
                    0,
                ],
                nodes: [
                    ConstantUnsignedInteger(
                        value: 82,
                        type_id: 1,
                    ),// [0]
                    Phi(
                        index: 0,
                    ),// [1]
                    Binop(
                        kind: Add,
                        arg0: 1,
                        arg1: 0,
                    ),// [2]
                    ConstantUnsignedInteger(
                        value: 81,
                        type_id: 1,
                    ),// [3]
                    Binop(
                        kind: Add,
                        arg0: 1,
                        arg1: 3,
                    ),// [4]
                    CallExternalCpu(
                        external_function_id: 0,
                        arguments: [
                            4,// [0]
                            2,
                        ],
                    ),
                ],
                tail_edge: Return(
                    return_values: [
                        5,
                    ],
                ),
                input_resource_states: [],
                output_resource_states: [],
                local_meta_variables: {},
            ),
            8: (
                kind: MixedImplicit,
                input_types: [
                    0,
                ],
                output_types: [
                    0,
                ],
                nodes: [
                    Phi(
                        index: 0,
                    ),
                ],
                tail_edge: Return(
                    return_values: [
                        0,
                    ],
                ),
                input_resource_states: [],
                output_resource_states: [],
                local_meta_variables: {},
            ),
            9: (
                kind: MixedImplicit,
                input_types: [
                    0,
                ],
                output_types: [
                    0,
                ],
                nodes: [
                    ConstantInteger(
                        value: 5,
                        type_id: 0,
                    ),// [0]
                    ConstantInteger(
                        value: -5,
                        type_id: 0,
                    ),// [1]
                    Phi(
                        index: 0,
                    ),// [2]
                    Binop(
                        kind: Add,
                        arg0: 2,
                        arg1: 1,
                    ),// [3]
                    Binop(
                        kind: Add,
                        arg0: 3,
                        arg1: 0,
                    ),
                ],
                tail_edge: Return(
                    return_values: [
                        4,
                    ],
                ),
                input_resource_states: [],
                output_resource_states: [],
                local_meta_variables: {},
            ),
        },
        external_cpu_functions: [],
        external_gpu_functions: [],
        value_functions: {},
        pipelines: [],
    ),
)